pkg/scope/scope.go

```diff
@@ -43,6 +43,21 @@ func NewFactory(maxCacheSize int) Factory {
 type Factory interface {
 	// NewClientScopeFromObject creates a new scope from the first object which returns an OpenStackIdentityRef
 	NewClientScopeFromObject(ctx context.Context, ctrlClient client.Client, defaultCACert []byte, logger logr.Logger, objects ...infrav1.IdentityRefProvider) (Scope, error)
+	GetIdentityRefFromObjects(objects ...infrav1.IdentityRefProvider) (*string, *infrav1.OpenStackIdentityReference)
+}
+
+type defaultScopeFactory struct{}
+
+func (f *defaultScopeFactory) GetIdentityRefFromObjects(objects ...infrav1.IdentityRefProvider) (*string, *infrav1.OpenStackIdentityReference) {
+	var namespace *string
+	var identityRef *infrav1.OpenStackIdentityReference
+	for _, o := range objects {
+		namespace, identityRef = o.GetIdentityRef()
+		if identityRef != nil {
+			break
+		}
+	}
+	return namespace, identityRef
```

Comment:
I'm not trying to be picky but I want to ensure we create code that will be maintainable and debuggable. I don't see error management here. What if the identifyRef wasn't found? Shouldn't we return an error?
Or from the callers, if identityRef is nil, we should return an error?

I think either way we need to stop and send a clean error if we can't find an identityRef. Especially because we use a pointer in one of the cases and I like to see pointer checks before using them.

Reply:
No problem, I ask myself about that too.
I took the choose of that the responsibility of caller since `objects ...infrav1.IdentityRefProvider` could have nil pointer to IdentityRef field, I don't wanna return an error because the value of objects could be nil as a normal behavior.

In my 2 callers functions last parameters are `openStackCluster` type of `infrav1.OpenStackCluster` which is not a pointer https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/api/v1beta1/openstackcluster_types.go#L194 so I'm sure that identityRef variable [here](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/2210/commits/dc72b3ce230ee1f5e05c7a56d91d93514dc57c60#diff-e0a0984c0e26655878e2f94623fbf8cddb4d8a89e3443c1296dfaadc8802838aR395) and [here](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/2210/commits/dc72b3ce230ee1f5e05c7a56d91d93514dc57c60#diff-e0a0984c0e26655878e2f94623fbf8cddb4d8a89e3443c1296dfaadc8802838aR579) and correctly defined.

But I'm agree maybe tomorrow it could be different, so I could validate function return from caller for more maintainability.